File: javascript/ql/test/library-tests/InterProceduralFlow/destructuring.js
(function() {
  var source = "tainted";
  let a = [source];
  let [tainted] = a;
  let sink = tainted;

  let b = { p: source };
  let { p: tainted2 } = b;
  let sink2 = tainted2;
})();

function f({ p, q: r }, o) {
  let { s } = o;
  let sink1 = p;
  let sink2 = r;
  let sink3 = s;
}

var source1 = "tainted";
var source2 = "also tainted";
var source3 = "still tainted";
f({ p: source1, q: source2 }, { s: source3 });

File: go/ql/test/library-tests/semmle/go/dataflow/InterProceduralDataFlow/destructuring.go
package main

import "fmt"

func test17() (string, string) {
	source1 := "tainted"
	return source1, "not tainted"
}

func test17b() (string, string) {
	source2 := "also tainted"
	return source2, "also not tainted"
}

func test17c() (string, string) {
	return test17b()
}

func test18(s1, s2 string) {
	sink1 := s1 // source1 flows here
	sink2 := s2
	fmt.Println(sink1, sink2)
}

func test19() {
	s1, s2 := test17()
	sink1 := s1 // source1 flows here
	sink2 := s2
	fmt.Println(sink1, sink2)

	test18(test17())

	s3, s4 := test17c()
	sink3 := s3 // source2 flows here
	sink4 := s4
	fmt.Println(sink3, sink4)
}
